core/client/db/src/lib.rs

```diff
@@ -892,9 +908,16 @@ impl<Block: BlockT<Hash=H256>> Backend<Block> {
 
 		if let Some((number, hash, enacted, retracted, displaced_leaf, is_best)) = imported {
 			if let Err(e) = write_result {
+				let mut leaves = self.blockchain.leaves.write();
```

Comment:
Could it be better to hold the lock for a little as possible, like in the below?

```
if let Some(displaced_leaf) = displaced_leaf {
					self.blockchain.leaves.write().undo().undo_import(displaced_leaf);
				}

				if let Some(finalization_displaced) = finalization_displaced_leaves {
					self.blockchain.leaves.write().undo().undo_finalization(finalization_displaced);
				}
```

Unless off-course you intend to exclude any reading or writing that might occur between the two method calls...

Reply:
`try_commit_operation` is meant to be synchronous as part of the API. It should only race with read calls, e.g. to `blockchain::Backend::leaves`. In that case, I think it's better either to give the leaves before, but not after.

Reply:
I'll file a broader issue about refactoring the `LeafSet` to use a transaction -> update pattern. https://github.com/paritytech/substrate/issues/1878